### PR TITLE
Convert hbflx3 to C++

### PR DESCRIPTION
**Pull request by maxpkatz**
## PR checklist

- [x] test suite needs to be run on this PR
- [ ] this PR will change answers in the test suite to more than roundoff level
- [ ] all newly-added functions have docstrings as per the coding conventions
- [ ] the `CHANGES` file has been updated, if appropriate
- [ ] if appropriate, this change is described in the docs

### CODE REVIEW

**Comment by zingale** [comment]
tests: http://groot.astro.sunysb.edu/Castro/test-suite/gfortran/2023-10-01/index.html
there are some FPEs

**Comment by maxpkatz** [comment]
> tests: http://groot.astro.sunysb.edu/Castro/test-suite/gfortran/2023-10-01/index.html there are some FPEs

I had an indexing error which is now fixed, can you try again?

File: Source/radiation/HABEC.H [hunk 1]
```diff
@@ -198,6 +198,486 @@ namespace HABEC
...
+                        }
+                        else if (bct == LO_NEUMANN) {
+                            bfv  = beta * r0;
+                            bfm  = 0.e0_rt;
+                            bfm2 = 0.e0_rt;
+                        }
+                        else if (bct == LO_MARSHAK) {
+                            bfv = 2.e0_rt * beta * r0;
+                            if (bho >= 1) {
+                                bfm  =  0.75e0_rt * beta * c;
```
Comment (zingale) [inline Source/radiation/HABEC.H:330]: these values are different than the 1-d case

File: Source/radiation/HABEC.H [hunk 2]
```diff
@@ -198,6 +198,486 @@ namespace HABEC
...
+                                bfm2 = -0.25e0_rt * beta * c;
+                            }
+                            else {
+                                bfm = 0.5e0_rt * beta * c;
+                            }
+                        }
+                        else if (bct == LO_SANCHEZ_POMRANING) {
+                            bfv = 2.e0_rt * beta * r0;
+                            if (bho >= 1) {
+                                bfm  =  3.0e0_rt * spa(i,j,k) * beta * c;
```
Comment (zingale) [inline Source/radiation/HABEC.H:340]: these expressions are different in the 1-d case

File: Source/radiation/HABEC.H [hunk 3]
```diff
@@ -198,6 +198,486 @@ namespace HABEC
...
+                        }
+                        else if (bct == LO_NEUMANN) {
+                            bfv  = beta * r0;
+                            bfm  = 0.e0_rt;
+                            bfm2 = 0.e0_rt;
+                        }
+                        else if (bct == LO_MARSHAK) {
+                            bfv = 2.e0_rt * beta * r0;
+                            if (bho >= 1) {
+                                bfm  =  0.75e0_rt * beta * c;
```
Comment (zingale) [inline Source/radiation/HABEC.H:393]: coefficients differ in 1-d

File: Source/radiation/HABEC.H [hunk 4]
```diff
@@ -198,6 +198,486 @@ namespace HABEC
...
+                                bfm2 = -0.25e0_rt * beta * c;
+                            }
+                            else {
+                                bfm = 0.5e0_rt * beta * c;
+                            }
+                        }
+                        else if (bct == LO_SANCHEZ_POMRANING) {
+                            bfv = 2.e0_rt * beta * r0;
```
Comment (zingale) [inline Source/radiation/HABEC.H:401]: bfv is not used?

File: Source/radiation/HABEC.H [hunk 5]
```diff
@@ -198,6 +198,486 @@ namespace HABEC
...
+                            if (bho >= 1) {
+                                bfm  =  3.0e0_rt * spa(i,j,k) * beta * c;
+                                bfm2 = -1.0e0_rt * spa(i,j,k) * beta * c;
```
Comment (zingale) [inline Source/radiation/HABEC.H:404]: coefficients differ in 1-d

File: Source/radiation/HABEC.H [hunk 6]
```diff
@@ -198,6 +198,486 @@ namespace HABEC
...
+                        }
+                        else if (bct == LO_NEUMANN) {
+                            bfv  = beta * r[i - reg_ilo];
+                            bfm  = 0.e0_rt;
+                            bfm2 = 0.e0_rt;
+                        }
+                        else if (bct == LO_MARSHAK) {
+                            bfv = 2.e0_rt * beta * r[i - reg_ilo];
+                            if (bho >= 1) {
+                                bfm  =  0.75e0_rt * beta * c;
```
Comment (zingale) [inline Source/radiation/HABEC.H:456]: coefficients differ in 2D

File: Source/radiation/HABEC.H [hunk 7]
```diff
@@ -198,6 +198,486 @@ namespace HABEC
...
+                                bfm2 = -0.25e0_rt * beta * c;
+                            }
+                            else {
+                                bfm = 0.5e0_rt * beta * c;
+                            }
+                        }
+                        else if (bct == LO_SANCHEZ_POMRANING) {
+                            bfv = 2.e0_rt * beta * r[i - reg_ilo];
+                            if (bho >= 1) {
+                                bfm  =  3.0e0_rt * spa(i,j,k) * beta * c;
```
Comment (zingale) [inline Source/radiation/HABEC.H:466]: coefficients differ in 2D

File: Source/radiation/HABEC.H [hunk 8]
```diff
@@ -198,6 +198,486 @@ namespace HABEC
...
```
Comment (zingale) [inline Source/radiation/HABEC.H:529]: coeffs differ in 2D

File: Source/radiation/HABEC.H [hunk 9]
```diff
@@ -198,6 +198,486 @@ namespace HABEC
...
+                            }
+                        }
+                        else if (bct == LO_NEUMANN) {
+                            bfv  = beta * r[i - reg_ilo];
+                            bfm  = 0.e0_rt;
+                            bfm2 = 0.e0_rt;
+                        }
+                        else if (bct == LO_MARSHAK) {
+                            bfv = 2.e0_rt * beta * r[i - reg_ilo];
+                            if (bho >= 1) {
```
Comment (zingale) [inline Source/radiation/HABEC.H:518]: coeffs differ in 2D

**Comment by maxpkatz** [comment]
I've synchronized 3D with 1D and 2D. The 3D case was making some simplifying assumptions; the more general expression in 1D and 2D are now what we use, and they result in the 3D expression because r0 = 1 for 3D Cartesian.

**Comment by zingale** [comment]
updated tests:
http://groot.astro.sunysb.edu/Castro/test-suite/gfortran/2023-10-07-004/index.html

File: Source/radiation/HABEC.H [hunk 10]
```diff
@@ -198,6 +198,486 @@ namespace HABEC
...
+                                bfm = bfv;
+                            }
+                        }
+                        else if (bct == LO_NEUMANN) {
+                            bfv  = beta * r[i - reg_ilo];
+                            bfm  = 0.e0_rt;
+                            bfm2 = 0.e0_rt;
+                        }
+                        else if (bct == LO_MARSHAK) {
+                            bfv = 2.e0_rt * beta * r[i - reg_ilo];
```
Comment (zingale) [inline Source/radiation/HABEC.H:454]: this works as expected in 3D?

**Comment by maxpkatz** [comment]
This PR now condenses all six faces of the grid into a single code block, and instead of using the vector of metric terms computed in a separate step, it now evaluates those metric terms inline.

**Comment by zingale** [comment]
all radiation tests pass:
http://groot.astro.sunysb.edu/Castro/test-suite/gfortran/2023-10-09-001/index.html

**Pull request merged**